### PR TITLE
snap-restore: added the config for temporarily disable check quorum

### PR DESCRIPTION
**Pull request by YuJuncen**
### What is changed and how it works?

Issue Number: Ref https://github.com/tikv/tikv/issues/15122

What's Changed:
This PR added a hidden, volatile config variable `restore.unsafe_skip_check_quorum`. This config term cannot be changed, and will only be modified by `init_cluster`.

```commit-message
```

### Check List 

Tests 

- Manual test (add detailed scripts or steps below)

### Release note 

(NOTE: the release note should be in https://github.com/pingcap/tidb/pull/45361)

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/config.rs [hunk 1]
```diff
@@ -107,6 +107,16 @@ pub struct Config {
...
 
+    /// Whether to disable checking quorum for the raft group. This will make
+    /// leader lease unavailable.
+    /// It cannot be changed in the config file,  the only way to change it is
+    /// programmatically change the config structure during bootstrapping
+    /// the cluster.
+    #[doc(hidden)]
+    #[serde(skip)]
+    #[online_config(skip)]
+    pub unsafe_skip_check_quorum: bool,
```
Comment (hicqu) [inline components/raftstore/src/store/config.rs:118]: `unsafe_disable_check_quorum` is better.

**Review by hicqu — APPROVED** [review]

File: components/snap_recovery/src/init_cluster.rs [hunk 2]
```diff
@@ -85,6 +85,14 @@ pub fn enter_snap_recovery_mode(config: &mut TikvConfig) {
     config.raft_store.max_snapshot_file_raw_size = ReadableSize::gb(1);
     config.raft_store.hibernate_regions = false;
 
+    // Disable prevote so it is possible to regenerate leaders.
+    config.raft_store.prevote = false;
+    // Due to we have increased the election tick to inf, once there is a leader,
```
Comment (tonyxuqqi) [inline components/snap_recovery/src/init_cluster.rs:90]: "Due to we have" ---> "Because we have".

File: components/raftstore/src/store/config.rs [hunk 3]
```diff
@@ -107,6 +107,16 @@ pub struct Config {
     #[online_config(skip)]
     pub raft_reject_transfer_leader_duration: ReadableDuration,
 
+    /// Whether to disable checking quorum for the raft group. This will make
+    /// leader lease unavailable.
+    /// It cannot be changed in the config file,  the only way to change it is
```
Comment (Connor1996) [inline components/raftstore/src/store/config.rs:112]:
```suggestion
    /// It cannot be changed in the config file, the only way to change it is
```

File: components/snap_recovery/src/init_cluster.rs [hunk 4]
```diff
@@ -85,6 +85,14 @@ pub fn enter_snap_recovery_mode(config: &mut TikvConfig) {
...
+    // the follower will believe it holds an eternal lease. So, once the leader
+    // reboots, the followers will reject to vote for it again.
```
Comment (Connor1996) [inline components/snap_recovery/src/init_cluster.rs:92]: where the code that followers reject to vote

Comment (tonyxuqqi) [reply components/snap_recovery/src/init_cluster.rs:92]: @Connor1996 Because election tick is inf so leader lease never expires.

**Comment by tonyxuqqi** [comment]
@YuJuncen What's the impact scope of this PR? Does it happen in restoring cluster? Will it impact after restore?

**Comment by YuJuncen** [comment]
> @YuJuncen What's the impact scope of this PR? Does it happen in restoring cluster? Will it impact after restore?

It won't impact after restore, because once snapshot restore has tweaked the cluster state to consistent, it reboots the cluster and the config flag won't be set any more.

File: components/raftstore/src/store/peer.rs [hunk 5]
```diff
@@ -825,7 +825,7 @@ where
             max_size_per_msg: cfg.raft_max_size_per_msg.0,
             max_inflight_msgs: cfg.raft_max_inflight_msgs,
             applied: applied_index,
-            check_quorum: true,
+            check_quorum: !cfg.unsafe_disable_check_quorum,
```
[the remark this answers is not in the document]
Comment (YuJuncen) [reply components/raftstore/src/store/peer.rs:828]: That is in raft-rs:

https://github.com/tikv/raft-rs/blob/9d360a3b0cdb691da8e500a4f73c457b605a1d73/src/raft.rs#L1328-L1331

**Review by BornChanger — APPROVED** [review]

**Review by tonyxuqqi — APPROVED** [review]

**Pull request merged**